### ISSUE TITLE
copyto! does not support CPU SubArrays

### ISSUE THREAD

**Issue by ww1g11** [JuliaGPU/GPUArrays.jl#422]

**Comment by david-macmahon** [comment]
I completely agree about not wanting to support all kinds of CPU arrays, but IMHO memory contiguous views seem like an important enough subset of CPU arrays to be worthy of support.

The use case where I first encountered this limitation was where I wanted to have B (where B >= 2) MxN matrices on the CPU so that when one is being transferred to the GPU the other(s) can still continue to be populated on the CPU side.  I first implemented this as a single MxNxB array on the CPU and made MxN views of each slice along the B dimension.  But I soon discovered that `copyto!` doesn't work with views (i.e. `SubArray`s).  The workaround for this case is to just make a vector (or tuple) of B separate MxN `Array`s.  This results in more calls to `Mem.pin` and more-but-smaller DMA mappings, which seems less efficient but haven't had any problems with that in practice.

A more problematic use case is where the GPU produces MxN parts of a larger MxBN output array.  I would like to be able to use `copyto!` to copy each MxN part directly into its place in the larger MxBN output array in CPU memory, but AFAIK this is not currently possible.  The workaround is to have B different MxN output arrays, but treating those as a larger MxBN array is cumbersome at best.

It seems like there must be some way to have a `copyto!` method that takes a `SubArray{T,N,Array,I,true}` and a `CuArray` and performs a DMA copy if the `SubArray` is memory contiguous or falls back to scalar indexing if the `SubArray` in not memory contiguous.  Maybe `SubArray{T,N,Array,I,true}` is already guaranteed to be memory contiguous?

**Comment by albertomercurio** [comment]
Hello,

do you have any news about this issue? I think that it would be very useful for a lot of applications.

**Comment by maleadt** [comment]
Nobody seems to be working on this, so no there have been no updates.
